src/charm.py

```diff
@@ -1,36 +1,163 @@
+        raft_storage_path=raft_storage_path,
+        node_id=node_id,
+    )
+    return content
+
+
+class Machine:
+    """A class to interact with a unit machine.
+
+    This class has the same method signatures as Pebble API in the Ops
+    Library. This is to improve consistency between the Machine and Kubernetes
+    versions of the charm.
+    """
+
+    def exists(self, path: str) -> bool:
+        """Check if a file exists.
+
+        Args:
+            path: The path of the file
+
+        Returns:
+            bool: Whether the file exists
+        """
+        return os.path.isfile(path)
+
+    def pull(self, path: str) -> str:
+        """Get the content of a file.
+
+        Args:
+            path: The path of the file
+
+        Returns:
+            str: The content of the file
+        """
+        with open(path, "r") as read_file:
+            return read_file.read()
+
+    def push(self, path: str, source: str) -> None:
+        """Pushes a file to the unit.
+
+        Args:
+            path: The path of the file
+            source: The contents of the file to be pushed
+        """
+        with open(path, "w") as write_file:
+            write_file.write(source)
+            logger.info("Pushed file %s", path)
+
+
+def config_file_content_matches(existing_content: str, new_content: str) -> bool:
+    """Return whether two Vault config file contents match.
+
+    We check if the retry_join addresses match, and then we check if the rest of the config
+    file matches.
+
+    Returns:
+        bool: Whether the vault config file content matches
+    """
+    existing_config_hcl = hcl.loads(existing_content)
+    new_content_hcl = hcl.loads(new_content)
+    if not existing_config_hcl:
+        logger.info("Existing config file is empty")
+        return existing_config_hcl == new_content_hcl
+    if not new_content_hcl:
+        logger.info("New config file is empty")
+        return existing_config_hcl == new_content_hcl
+
+    new_retry_joins = new_content_hcl["storage"]["raft"].pop("retry_join", [])
+    existing_retry_joins = existing_config_hcl["storage"]["raft"].pop("retry_join", [])
+
+    # If there is only one retry join, it is a dict
+    if isinstance(new_retry_joins, dict):
+        new_retry_joins = [new_retry_joins]
+    if isinstance(existing_retry_joins, dict):
+        existing_retry_joins = [existing_retry_joins]
+
+    new_retry_join_api_addresses = {address["leader_api_addr"] for address in new_retry_joins}
+    existing_retry_join_api_addresses = {
+        address["leader_api_addr"] for address in existing_retry_joins
+    }
+    return (
+        new_retry_join_api_addresses == existing_retry_join_api_addresses
+        and new_content_hcl == existing_config_hcl
+    )
 
 
 class VaultOperatorCharm(CharmBase):
-    """Charm the service."""
+    """Machine Charm for Vault."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self.unit_file_directory = Machine()
```

Comment:
How about this name instead?
```suggestion
        self.machine = Machine()
```

Reply:
Done